{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn import preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load training dataset\n",
    "with open('datasets/train.dataset.6mer.npy', 'rb') as open_file:\n",
    "    df = np.load(open_file)\n",
    "df = pd.DataFrame(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = pd.read_csv('datasets/train_labels.csv')\n",
    "\n",
    "# snippet to load the grouth truth training labels and normalize the label predictions. \n",
    "# your trained model will predict in this space (31 classes - pathogens and decoy)\n",
    "le = preprocessing.LabelEncoder()\n",
    "le.fit(labels['genome_name'].unique())\n",
    "y_index = le.transform(labels['genome_name'].values)\n",
    "\n",
    "# use index as train labels\n",
    "df['genome_label'] = y_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7     424388\n",
       "1       3787\n",
       "16      3342\n",
       "8       3166\n",
       "13      2999\n",
       "9       2840\n",
       "19      2832\n",
       "3       2715\n",
       "18      2594\n",
       "17      2578\n",
       "30      2404\n",
       "23      2388\n",
       "12      2354\n",
       "4       2248\n",
       "0       2124\n",
       "10      1801\n",
       "29      1738\n",
       "11      1576\n",
       "28      1561\n",
       "20      1489\n",
       "21      1376\n",
       "6       1306\n",
       "5       1274\n",
       "15      1194\n",
       "25      1182\n",
       "14      1150\n",
       "27      1128\n",
       "24      1092\n",
       "26      1047\n",
       "22       942\n",
       "2        910\n",
       "Name: genome_label, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['genome_label'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>22</th>\n",
       "      <th>...</th>\n",
       "      <th>2071</th>\n",
       "      <th>2072</th>\n",
       "      <th>2073</th>\n",
       "      <th>2074</th>\n",
       "      <th>2075</th>\n",
       "      <th>2076</th>\n",
       "      <th>2077</th>\n",
       "      <th>2078</th>\n",
       "      <th>2079</th>\n",
       "      <th>genome_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000823</td>\n",
       "      <td>0.001646</td>\n",
       "      <td>0.000823</td>\n",
       "      <td>0.002470</td>\n",
       "      <td>0.000411</td>\n",
       "      <td>0.001235</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001235</td>\n",
       "      <td>0.001235</td>\n",
       "      <td>0.000411</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.001558</td>\n",
       "      <td>0.001039</td>\n",
       "      <td>0.001299</td>\n",
       "      <td>0.000779</td>\n",
       "      <td>0.000779</td>\n",
       "      <td>0.000779</td>\n",
       "      <td>0.000779</td>\n",
       "      <td>0.000779</td>\n",
       "      <td>0.002077</td>\n",
       "      <td>0.002077</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000519</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000519</td>\n",
       "      <td>0.000260</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.001851</td>\n",
       "      <td>0.000617</td>\n",
       "      <td>0.001234</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000617</td>\n",
       "      <td>0.001234</td>\n",
       "      <td>0.001851</td>\n",
       "      <td>0.002468</td>\n",
       "      <td>0.000617</td>\n",
       "      <td>0.001851</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.001378</td>\n",
       "      <td>0.001378</td>\n",
       "      <td>0.000276</td>\n",
       "      <td>0.003859</td>\n",
       "      <td>0.001102</td>\n",
       "      <td>0.001102</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001378</td>\n",
       "      <td>0.001102</td>\n",
       "      <td>0.002756</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.001701</td>\n",
       "      <td>0.001440</td>\n",
       "      <td>0.001178</td>\n",
       "      <td>0.001047</td>\n",
       "      <td>0.001047</td>\n",
       "      <td>0.001178</td>\n",
       "      <td>0.000785</td>\n",
       "      <td>0.000916</td>\n",
       "      <td>0.002487</td>\n",
       "      <td>0.001571</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000262</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1919 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          3         9        10        11        13        14        15  \\\n",
       "0  0.000823  0.001646  0.000823  0.002470  0.000411  0.001235  0.000000   \n",
       "1  0.001558  0.001039  0.001299  0.000779  0.000779  0.000779  0.000779   \n",
       "2  0.001851  0.000617  0.001234  0.000000  0.000617  0.001234  0.001851   \n",
       "3  0.001378  0.001378  0.000276  0.003859  0.001102  0.001102  0.000000   \n",
       "4  0.001701  0.001440  0.001178  0.001047  0.001047  0.001178  0.000785   \n",
       "\n",
       "         18        19        22  ...      2071  2072  2073      2074  \\\n",
       "0  0.001235  0.001235  0.000411  ...  0.000000   0.0   0.0  0.000000   \n",
       "1  0.000779  0.002077  0.002077  ...  0.000519   0.0   0.0  0.000519   \n",
       "2  0.002468  0.000617  0.001851  ...  0.000000   0.0   0.0  0.000000   \n",
       "3  0.001378  0.001102  0.002756  ...  0.000000   0.0   0.0  0.000000   \n",
       "4  0.000916  0.002487  0.001571  ...  0.000000   0.0   0.0  0.000262   \n",
       "\n",
       "       2075  2076  2077  2078  2079  genome_label  \n",
       "0  0.000000   0.0   0.0   0.0   0.0            20  \n",
       "1  0.000260   0.0   0.0   0.0   0.0            20  \n",
       "2  0.000000   0.0   0.0   0.0   0.0            20  \n",
       "3  0.000000   0.0   0.0   0.0   0.0            20  \n",
       "4  0.000131   0.0   0.0   0.0   0.0            20  \n",
       "\n",
       "[5 rows x 1919 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(62924, 1919)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove samples with total k-mer count < 0.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_kmers = df.iloc[:, :-1].sum(axis=1)\n",
    "df = df.loc[total_kmers >= 0.9, :]\n",
    "y_index = y_index[total_kmers >= 0.9]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Downsample decoys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([415385, 276717, 301096,  88285,  15499, 377950, 329334, 425205,\n",
       "             90461, 477715,\n",
       "            ...\n",
       "             94812, 381037, 167116, 117181, 139215, 154100, 240125, 385251,\n",
       "            431883, 105292],\n",
       "           dtype='int64', length=420601)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "downsampled = df[df['genome_label'] == 7].sample(420601, random_state=4)\n",
    "downsampled.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop(index=downsampled.index, inplace=True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove columns with high correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = np.triu(np.ones_like(df.corr(), dtype=bool)) # remove duplicate upper triangle correlation values\n",
    "corr_matrix = df.corr().abs() # create positive correlation matrix\n",
    "tri_df = corr_matrix.mask(mask) # create and apply upper triangle mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "162"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "high_corr = [c for c in tri_df.columns if any(tri_df[c] > 0.7)]\n",
    "len(high_corr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'[0 1 2 4 5 6 7 8 12 16 17 20 21 24 28 32 36 48 64 65 68 69 80 81 84 127\\n 189 250 251 254 265 266 267 269 280 310 311 314 325 326 328 427 441 465\\n 468 484 513 644 647 658 684 687 688 743 813 815 827 840 850 851 853 888\\n 890 891 904 905 907 916 952 987 999 1090 1091 1112 1139 1147 1148 1149\\n 1152 1155 1246 1248 1373 1375 1391 1392 1393 1403 1404 1411 1490 1491\\n 1499 1508 1509 1514 1515 1516 1517 1605 1612 1642 1643 1664 1665 1666\\n 1671 1746 1752 1771 1772 1818 1820 1845 1846 1847 1852 1858 1859 1861\\n 1862 1863 1864 1865 1866 1867 1868 1869 1883 1884 1888 1889 1902 1903\\n 1906 1907 1912 1916 1918 1920 1921 1922 1923 1924 1925 1926 1929 1934\\n 1935 1937 1938 1939 1987 1998 2000 2031 2041 2049 2050 2056 2062 2063] not found in axis'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/61/_1tynjrd31972whvq0rhfvmc0000gn/T/ipykernel_10910/809209129.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf_red\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mhigh_corr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/util/_decorators.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    309\u001b[0m                     \u001b[0mstacklevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstacklevel\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    310\u001b[0m                 )\n\u001b[0;32m--> 311\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    313\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mdrop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   4904\u001b[0m                 \u001b[0mweight\u001b[0m  \u001b[0;36m1.0\u001b[0m     \u001b[0;36m0.8\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4905\u001b[0m         \"\"\"\n\u001b[0;32m-> 4906\u001b[0;31m         return super().drop(\n\u001b[0m\u001b[1;32m   4907\u001b[0m             \u001b[0mlabels\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4908\u001b[0m             \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mdrop\u001b[0;34m(self, labels, axis, index, columns, level, inplace, errors)\u001b[0m\n\u001b[1;32m   4148\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels\u001b[0m \u001b[0;32min\u001b[0m \u001b[0maxes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4149\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mlabels\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4150\u001b[0;31m                 \u001b[0mobj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_drop_axis\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4151\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4152\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_drop_axis\u001b[0;34m(self, labels, axis, level, errors)\u001b[0m\n\u001b[1;32m   4183\u001b[0m                 \u001b[0mnew_axis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mlevel\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4184\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4185\u001b[0;31m                 \u001b[0mnew_axis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merrors\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merrors\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4186\u001b[0m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0maxis_name\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mnew_axis\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4187\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mdrop\u001b[0;34m(self, labels, errors)\u001b[0m\n\u001b[1;32m   6015\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6016\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0merrors\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m\"ignore\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6017\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{labels[mask]} not found in axis\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6018\u001b[0m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0mmask\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6019\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdelete\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindexer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: '[0 1 2 4 5 6 7 8 12 16 17 20 21 24 28 32 36 48 64 65 68 69 80 81 84 127\\n 189 250 251 254 265 266 267 269 280 310 311 314 325 326 328 427 441 465\\n 468 484 513 644 647 658 684 687 688 743 813 815 827 840 850 851 853 888\\n 890 891 904 905 907 916 952 987 999 1090 1091 1112 1139 1147 1148 1149\\n 1152 1155 1246 1248 1373 1375 1391 1392 1393 1403 1404 1411 1490 1491\\n 1499 1508 1509 1514 1515 1516 1517 1605 1612 1642 1643 1664 1665 1666\\n 1671 1746 1752 1771 1772 1818 1820 1845 1846 1847 1852 1858 1859 1861\\n 1862 1863 1864 1865 1866 1867 1868 1869 1883 1884 1888 1889 1902 1903\\n 1906 1907 1912 1916 1918 1920 1921 1922 1923 1924 1925 1926 1929 1934\\n 1935 1937 1938 1939 1987 1998 2000 2031 2041 2049 2050 2056 2062 2063] not found in axis'"
     ]
    }
   ],
   "source": [
    "df.drop(columns=high_corr, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('pre_svd.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('pre_svd.csv')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Feature extraction using tSVD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TruncatedSVD(n_components=500, random_state=4)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.decomposition import TruncatedSVD\n",
    "svd = TruncatedSVD(n_components=500, random_state=4)\n",
    "svd.fit(df.iloc[:, :-1].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle as pk\n",
    "pk.dump(svd, open('transformers/svd.pkl', 'wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle as pk\n",
    "svd = pk.load(open('transformers/svd.pkl', 'rb'))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Subsample training set and perform dimensionality reduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_train = df.groupby('genome_label').apply(lambda x: x.sample(910, random_state=4))\n",
    "X_train = svd.transform(df_train.iloc[:, :-1].values)\n",
    "y_train = df_train['genome_label'].astype(int).values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 2.16119806e-02, -1.12246967e-02,  3.10852930e-04, ...,\n",
       "         1.48419513e-04, -3.78807179e-04, -2.21491643e-04],\n",
       "       [ 2.18505849e-02, -1.15034471e-02,  1.63952861e-03, ...,\n",
       "         2.48144278e-04,  6.19929214e-04, -2.66408205e-04],\n",
       "       [ 2.15662925e-02, -9.79539106e-03,  7.32436260e-04, ...,\n",
       "         3.93659798e-04, -9.01028979e-05, -6.77395971e-05],\n",
       "       ...,\n",
       "       [ 2.17961952e-02, -5.53790644e-03, -2.73646639e-03, ...,\n",
       "         5.97028994e-04, -4.89305601e-05,  1.07936952e-04],\n",
       "       [ 2.26182567e-02, -1.38002959e-03, -4.36690458e-03, ...,\n",
       "        -2.82249076e-04, -1.28196469e-05, -3.08790280e-05],\n",
       "       [ 2.21711476e-02, -2.21356876e-03, -4.22842009e-03, ...,\n",
       "         2.35392374e-04, -5.13989618e-04, -2.91539009e-04]])"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>3</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>22</th>\n",
       "      <th>...</th>\n",
       "      <th>2071</th>\n",
       "      <th>2072</th>\n",
       "      <th>2073</th>\n",
       "      <th>2074</th>\n",
       "      <th>2075</th>\n",
       "      <th>2076</th>\n",
       "      <th>2077</th>\n",
       "      <th>2078</th>\n",
       "      <th>2079</th>\n",
       "      <th>genome_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genome_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">0</th>\n",
       "      <th>34420</th>\n",
       "      <td>0.003378</td>\n",
       "      <td>0.001987</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.002783</td>\n",
       "      <td>0.001590</td>\n",
       "      <td>0.001391</td>\n",
       "      <td>0.001391</td>\n",
       "      <td>0.001391</td>\n",
       "      <td>0.001391</td>\n",
       "      <td>0.000795</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000795</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000596</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32679</th>\n",
       "      <td>0.001767</td>\n",
       "      <td>0.001632</td>\n",
       "      <td>0.000816</td>\n",
       "      <td>0.001224</td>\n",
       "      <td>0.001767</td>\n",
       "      <td>0.000544</td>\n",
       "      <td>0.001495</td>\n",
       "      <td>0.001359</td>\n",
       "      <td>0.000544</td>\n",
       "      <td>0.001495</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000136</td>\n",
       "      <td>0.000272</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33660</th>\n",
       "      <td>0.001410</td>\n",
       "      <td>0.000846</td>\n",
       "      <td>0.000564</td>\n",
       "      <td>0.001974</td>\n",
       "      <td>0.002115</td>\n",
       "      <td>0.001410</td>\n",
       "      <td>0.001268</td>\n",
       "      <td>0.000564</td>\n",
       "      <td>0.000846</td>\n",
       "      <td>0.000987</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000282</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000141</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34726</th>\n",
       "      <td>0.001965</td>\n",
       "      <td>0.002321</td>\n",
       "      <td>0.000535</td>\n",
       "      <td>0.001071</td>\n",
       "      <td>0.002142</td>\n",
       "      <td>0.000893</td>\n",
       "      <td>0.001965</td>\n",
       "      <td>0.000893</td>\n",
       "      <td>0.001785</td>\n",
       "      <td>0.001250</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000535</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000357</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32832</th>\n",
       "      <td>0.000804</td>\n",
       "      <td>0.001608</td>\n",
       "      <td>0.000804</td>\n",
       "      <td>0.002813</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001205</td>\n",
       "      <td>0.002010</td>\n",
       "      <td>0.000804</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002010</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000804</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000402</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">30</th>\n",
       "      <th>57745</th>\n",
       "      <td>0.001081</td>\n",
       "      <td>0.000480</td>\n",
       "      <td>0.000360</td>\n",
       "      <td>0.000480</td>\n",
       "      <td>0.000960</td>\n",
       "      <td>0.000480</td>\n",
       "      <td>0.000840</td>\n",
       "      <td>0.001560</td>\n",
       "      <td>0.000840</td>\n",
       "      <td>0.000480</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000240</td>\n",
       "      <td>0.000480</td>\n",
       "      <td>0.000840</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000480</td>\n",
       "      <td>0.000360</td>\n",
       "      <td>0.000240</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57330</th>\n",
       "      <td>0.000887</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.000380</td>\n",
       "      <td>0.000634</td>\n",
       "      <td>0.000760</td>\n",
       "      <td>0.000380</td>\n",
       "      <td>0.000380</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.000760</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.000634</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000507</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.000127</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58178</th>\n",
       "      <td>0.000985</td>\n",
       "      <td>0.001232</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000739</td>\n",
       "      <td>0.000246</td>\n",
       "      <td>0.000493</td>\n",
       "      <td>0.000739</td>\n",
       "      <td>0.000739</td>\n",
       "      <td>0.000985</td>\n",
       "      <td>0.000739</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000246</td>\n",
       "      <td>0.000246</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000246</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000493</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57039</th>\n",
       "      <td>0.001042</td>\n",
       "      <td>0.000651</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.000521</td>\n",
       "      <td>0.000521</td>\n",
       "      <td>0.000782</td>\n",
       "      <td>0.001564</td>\n",
       "      <td>0.000782</td>\n",
       "      <td>0.000782</td>\n",
       "      <td>0.000782</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000521</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.001173</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.000391</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57340</th>\n",
       "      <td>0.001428</td>\n",
       "      <td>0.000408</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001224</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.000816</td>\n",
       "      <td>0.000612</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.000816</td>\n",
       "      <td>0.000408</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000408</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.000408</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.000408</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28210 rows × 1919 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           3         9        10        11        13  \\\n",
       "genome_label                                                           \n",
       "0            34420  0.003378  0.001987  0.000199  0.002783  0.001590   \n",
       "             32679  0.001767  0.001632  0.000816  0.001224  0.001767   \n",
       "             33660  0.001410  0.000846  0.000564  0.001974  0.002115   \n",
       "             34726  0.001965  0.002321  0.000535  0.001071  0.002142   \n",
       "             32832  0.000804  0.001608  0.000804  0.002813  0.000000   \n",
       "...                      ...       ...       ...       ...       ...   \n",
       "30           57745  0.001081  0.000480  0.000360  0.000480  0.000960   \n",
       "             57330  0.000887  0.000127  0.000253  0.000380  0.000634   \n",
       "             58178  0.000985  0.001232  0.000000  0.000739  0.000246   \n",
       "             57039  0.001042  0.000651  0.000261  0.000521  0.000521   \n",
       "             57340  0.001428  0.000408  0.000000  0.001224  0.000204   \n",
       "\n",
       "                          14        15        18        19        22  ...  \\\n",
       "genome_label                                                          ...   \n",
       "0            34420  0.001391  0.001391  0.001391  0.001391  0.000795  ...   \n",
       "             32679  0.000544  0.001495  0.001359  0.000544  0.001495  ...   \n",
       "             33660  0.001410  0.001268  0.000564  0.000846  0.000987  ...   \n",
       "             34726  0.000893  0.001965  0.000893  0.001785  0.001250  ...   \n",
       "             32832  0.001205  0.002010  0.000804  0.000000  0.002010  ...   \n",
       "...                      ...       ...       ...       ...       ...  ...   \n",
       "30           57745  0.000480  0.000840  0.001560  0.000840  0.000480  ...   \n",
       "             57330  0.000760  0.000380  0.000380  0.000253  0.000760  ...   \n",
       "             58178  0.000493  0.000739  0.000739  0.000985  0.000739  ...   \n",
       "             57039  0.000782  0.001564  0.000782  0.000782  0.000782  ...   \n",
       "             57340  0.000816  0.000612  0.000204  0.000816  0.000408  ...   \n",
       "\n",
       "                        2071      2072      2073      2074      2075  \\\n",
       "genome_label                                                           \n",
       "0            34420  0.000795  0.000000  0.000596  0.000000  0.000000   \n",
       "             32679  0.000000  0.000136  0.000272  0.000000  0.000000   \n",
       "             33660  0.000000  0.000000  0.000282  0.000000  0.000141   \n",
       "             34726  0.000000  0.000000  0.000535  0.000000  0.000000   \n",
       "             32832  0.000804  0.000000  0.000402  0.000000  0.000000   \n",
       "...                      ...       ...       ...       ...       ...   \n",
       "30           57745  0.000240  0.000480  0.000840  0.000000  0.000480   \n",
       "             57330  0.000253  0.000127  0.000634  0.000000  0.000507   \n",
       "             58178  0.000000  0.000246  0.000246  0.000000  0.000246   \n",
       "             57039  0.000521  0.000130  0.001173  0.000261  0.000391   \n",
       "             57340  0.000000  0.000000  0.000408  0.000000  0.000204   \n",
       "\n",
       "                        2076      2077      2078  2079  genome_label  \n",
       "genome_label                                                          \n",
       "0            34420  0.000000  0.000000  0.000000   0.0             0  \n",
       "             32679  0.000000  0.000000  0.000000   0.0             0  \n",
       "             33660  0.000000  0.000000  0.000000   0.0             0  \n",
       "             34726  0.000000  0.000000  0.000357   0.0             0  \n",
       "             32832  0.000000  0.000000  0.000000   0.0             0  \n",
       "...                      ...       ...       ...   ...           ...  \n",
       "30           57745  0.000360  0.000240  0.000000   0.0            30  \n",
       "             57330  0.000127  0.000127  0.000000   0.0            30  \n",
       "             58178  0.000000  0.000000  0.000493   0.0            30  \n",
       "             57039  0.000130  0.000261  0.000000   0.0            30  \n",
       "             57340  0.000408  0.000204  0.000408   0.0            30  \n",
       "\n",
       "[28210 rows x 1919 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Train SVM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CalibratedClassifierCV(base_estimator=SVC(C=10, gamma=10, random_state=4), cv=5)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sklearn.calibration import CalibratedClassifierCV\n",
    "from sklearn.svm import SVC\n",
    "from joblib import dump, load\n",
    "svm = SVC(C=10, kernel='rbf', gamma=10, probability=False, random_state=4)\n",
    "model = CalibratedClassifierCV(svm, method='sigmoid', cv=5) # wrap the SVM classifier in a CalibratedClassifierCV instance\n",
    "model.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['models/svm.joblib']"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dump(model, 'models/svm.joblib')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "from joblib import load\n",
    "model = load('models/svm.joblib')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def precision_per_patient(patient_id, preds):\n",
    "    df_true = pd.read_csv(f'datasets/validation/patient{patient_id}_labels.txt')\n",
    "    tp, fp, tp_labels = 0, 0, df_true['true_label'].shape[0]\n",
    "    print(f'my prediction(s) for patient {patient_id}: {preds}')\n",
    "    truth = df_true['true_label'].values\n",
    "    print(f'true pathogen(s): {truth}')\n",
    "    #if don't predict any pathogen, it means there is only decoy in the test dataset (your prediction)\n",
    "    if len(preds) == 0:\n",
    "        preds = ['decoy']\n",
    "    for item in np.unique(preds):\n",
    "        if item in df_true['true_label'].values:\n",
    "            tp+=1\n",
    "        else:\n",
    "            fp+=1\n",
    "    #you have to predict all labels correctly, but you are penalized for any false positive\n",
    "    return tp/(tp_labels+fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>2070</th>\n",
       "      <th>2071</th>\n",
       "      <th>2072</th>\n",
       "      <th>2073</th>\n",
       "      <th>2074</th>\n",
       "      <th>2075</th>\n",
       "      <th>2076</th>\n",
       "      <th>2077</th>\n",
       "      <th>2078</th>\n",
       "      <th>2079</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000655</td>\n",
       "      <td>0.003929</td>\n",
       "      <td>0.001309</td>\n",
       "      <td>0.001528</td>\n",
       "      <td>0.002401</td>\n",
       "      <td>0.003492</td>\n",
       "      <td>0.001965</td>\n",
       "      <td>0.000655</td>\n",
       "      <td>0.000873</td>\n",
       "      <td>0.001091</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000218</td>\n",
       "      <td>0.000873</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000437</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.001521</td>\n",
       "      <td>0.000869</td>\n",
       "      <td>0.003042</td>\n",
       "      <td>0.001304</td>\n",
       "      <td>0.001304</td>\n",
       "      <td>0.000652</td>\n",
       "      <td>0.000652</td>\n",
       "      <td>0.001521</td>\n",
       "      <td>0.002172</td>\n",
       "      <td>0.001304</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000217</td>\n",
       "      <td>0.000434</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.001271</td>\n",
       "      <td>0.003496</td>\n",
       "      <td>0.000636</td>\n",
       "      <td>0.001271</td>\n",
       "      <td>0.002224</td>\n",
       "      <td>0.002542</td>\n",
       "      <td>0.003813</td>\n",
       "      <td>0.000953</td>\n",
       "      <td>0.001271</td>\n",
       "      <td>0.000953</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000636</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.002781</td>\n",
       "      <td>0.002411</td>\n",
       "      <td>0.001483</td>\n",
       "      <td>0.000927</td>\n",
       "      <td>0.002781</td>\n",
       "      <td>0.001854</td>\n",
       "      <td>0.001669</td>\n",
       "      <td>0.000556</td>\n",
       "      <td>0.002039</td>\n",
       "      <td>0.000741</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000741</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000371</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.001933</td>\n",
       "      <td>0.003410</td>\n",
       "      <td>0.001478</td>\n",
       "      <td>0.001592</td>\n",
       "      <td>0.003639</td>\n",
       "      <td>0.003298</td>\n",
       "      <td>0.003071</td>\n",
       "      <td>0.001592</td>\n",
       "      <td>0.001137</td>\n",
       "      <td>0.001250</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.000227</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9993</th>\n",
       "      <td>0.001433</td>\n",
       "      <td>0.002867</td>\n",
       "      <td>0.001433</td>\n",
       "      <td>0.000358</td>\n",
       "      <td>0.002867</td>\n",
       "      <td>0.003225</td>\n",
       "      <td>0.002867</td>\n",
       "      <td>0.001433</td>\n",
       "      <td>0.002151</td>\n",
       "      <td>0.000358</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000358</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.001076</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000717</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9994</th>\n",
       "      <td>0.008995</td>\n",
       "      <td>0.003229</td>\n",
       "      <td>0.001384</td>\n",
       "      <td>0.000923</td>\n",
       "      <td>0.003229</td>\n",
       "      <td>0.001384</td>\n",
       "      <td>0.001615</td>\n",
       "      <td>0.001153</td>\n",
       "      <td>0.000461</td>\n",
       "      <td>0.001384</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000692</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000461</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000461</td>\n",
       "      <td>0.001153</td>\n",
       "      <td>0.000231</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000461</td>\n",
       "      <td>0.000461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9995</th>\n",
       "      <td>0.004356</td>\n",
       "      <td>0.001452</td>\n",
       "      <td>0.001816</td>\n",
       "      <td>0.002905</td>\n",
       "      <td>0.002542</td>\n",
       "      <td>0.001089</td>\n",
       "      <td>0.003267</td>\n",
       "      <td>0.000363</td>\n",
       "      <td>0.001452</td>\n",
       "      <td>0.001452</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000363</td>\n",
       "      <td>0.000726</td>\n",
       "      <td>0.000726</td>\n",
       "      <td>0.000363</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000726</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9996</th>\n",
       "      <td>0.000876</td>\n",
       "      <td>0.000876</td>\n",
       "      <td>0.002628</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001753</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000876</td>\n",
       "      <td>0.001753</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003506</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000876</td>\n",
       "      <td>0.003506</td>\n",
       "      <td>0.001753</td>\n",
       "      <td>0.000876</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001753</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9997</th>\n",
       "      <td>0.004841</td>\n",
       "      <td>0.001760</td>\n",
       "      <td>0.003080</td>\n",
       "      <td>0.001321</td>\n",
       "      <td>0.003521</td>\n",
       "      <td>0.002201</td>\n",
       "      <td>0.002201</td>\n",
       "      <td>0.001321</td>\n",
       "      <td>0.002201</td>\n",
       "      <td>0.001760</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000440</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000880</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9998 rows × 2080 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2         3         4         5         6     \\\n",
       "0     0.000655  0.003929  0.001309  0.001528  0.002401  0.003492  0.001965   \n",
       "1     0.001521  0.000869  0.003042  0.001304  0.001304  0.000652  0.000652   \n",
       "2     0.001271  0.003496  0.000636  0.001271  0.002224  0.002542  0.003813   \n",
       "3     0.002781  0.002411  0.001483  0.000927  0.002781  0.001854  0.001669   \n",
       "4     0.001933  0.003410  0.001478  0.001592  0.003639  0.003298  0.003071   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "9993  0.001433  0.002867  0.001433  0.000358  0.002867  0.003225  0.002867   \n",
       "9994  0.008995  0.003229  0.001384  0.000923  0.003229  0.001384  0.001615   \n",
       "9995  0.004356  0.001452  0.001816  0.002905  0.002542  0.001089  0.003267   \n",
       "9996  0.000876  0.000876  0.002628  0.000000  0.001753  0.000000  0.000876   \n",
       "9997  0.004841  0.001760  0.003080  0.001321  0.003521  0.002201  0.002201   \n",
       "\n",
       "          7         8         9     ...      2070      2071      2072  \\\n",
       "0     0.000655  0.000873  0.001091  ...  0.000218  0.000873  0.000000   \n",
       "1     0.001521  0.002172  0.001304  ...  0.000000  0.000000  0.000000   \n",
       "2     0.000953  0.001271  0.000953  ...  0.000000  0.000636  0.000000   \n",
       "3     0.000556  0.002039  0.000741  ...  0.000000  0.000741  0.000000   \n",
       "4     0.001592  0.001137  0.001250  ...  0.000114  0.000227  0.000000   \n",
       "...        ...       ...       ...  ...       ...       ...       ...   \n",
       "9993  0.001433  0.002151  0.000358  ...  0.000358  0.000717  0.001076   \n",
       "9994  0.001153  0.000461  0.001384  ...  0.000692  0.000000  0.000461   \n",
       "9995  0.000363  0.001452  0.001452  ...  0.000363  0.000726  0.000726   \n",
       "9996  0.001753  0.000000  0.003506  ...  0.000876  0.003506  0.001753   \n",
       "9997  0.001321  0.002201  0.001760  ...  0.000440  0.000000  0.000000   \n",
       "\n",
       "          2073      2074      2075      2076  2077      2078      2079  \n",
       "0     0.000000  0.000437  0.000000  0.000000   0.0  0.000000  0.000000  \n",
       "1     0.000217  0.000434  0.000000  0.000000   0.0  0.000000  0.000000  \n",
       "2     0.000000  0.000000  0.000000  0.000000   0.0  0.000000  0.000000  \n",
       "3     0.000000  0.000000  0.000371  0.000000   0.0  0.000000  0.000000  \n",
       "4     0.000114  0.000000  0.000000  0.000000   0.0  0.000000  0.000000  \n",
       "...        ...       ...       ...       ...   ...       ...       ...  \n",
       "9993  0.000000  0.000717  0.000000  0.000000   0.0  0.000000  0.000000  \n",
       "9994  0.000000  0.000461  0.001153  0.000231   0.0  0.000461  0.000461  \n",
       "9995  0.000363  0.000000  0.000726  0.000000   0.0  0.000000  0.000000  \n",
       "9996  0.000876  0.000000  0.001753  0.000000   0.0  0.000000  0.000000  \n",
       "9997  0.000000  0.000880  0.000000  0.000000   0.0  0.000000  0.000000  \n",
       "\n",
       "[9998 rows x 2080 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(f'datasets/validation/patient1.6mer.npy', 'rb') as read_file:\n",
    "    df_test = np.load(read_file)\n",
    "df_test = pd.DataFrame(df_test)\n",
    "df_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "feat = df_train.columns[:-1].values.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1918"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(feat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "predicting for patient 1\n",
      "{1: 52, 5: 6, 16: 4, 11: 64, 7: 2445}\n",
      "my prediction(s) for patient 1: ['burkholderia_pseudomallei', 'corynebacterium_diphtheriae', 'listeria_monocytogenes']\n",
      "true pathogen(s): ['staphylococcus_aureus']\n",
      "precision: 0.0\n",
      "predicting for patient 2\n",
      "{22: 39, 29: 6, 25: 4, 6: 9, 0: 4, 5: 72, 4: 3, 11: 455, 21: 5, 20: 6, 10: 8, 1: 482, 16: 5, 13: 15, 27: 3, 26: 1, 24: 1, 7: 881}\n",
      "my prediction(s) for patient 2: ['staphylococcus_pyogenes', 'vibrio_parahaemolyticus', 'corynebacterium_ulcerans', 'corynebacterium_diphtheriae', 'listeria_monocytogenes', 'staphylococcus_aureus', 'legionella_pneumophila', 'burkholderia_pseudomallei', 'mycobacterium_ulcerans']\n",
      "true pathogen(s): ['staphylococcus_pyogenes']\n",
      "precision: 0.1111111111111111\n",
      "predicting for patient 3\n",
      "{1: 136, 5: 17, 16: 3, 13: 3, 6: 8, 11: 867, 10: 11, 22: 7, 0: 5, 21: 7, 27: 3, 26: 3, 24: 3, 7: 1392}\n",
      "my prediction(s) for patient 3: ['burkholderia_pseudomallei', 'corynebacterium_diphtheriae', 'corynebacterium_ulcerans', 'listeria_monocytogenes', 'legionella_pneumophila', 'staphylococcus_pyogenes', 'staphylococcus_pseudintermedius']\n",
      "true pathogen(s): ['burkholderia_pseudomallei' 'corynebacterium_ulcerans']\n",
      "precision: 0.2857142857142857\n",
      "predicting for patient 4\n",
      "{16: 7, 5: 2, 13: 15, 6: 1, 11: 241, 22: 1, 10: 7, 27: 3, 21: 2, 30: 1, 20: 2, 29: 1, 7: 1131}\n",
      "my prediction(s) for patient 4: ['pseudomonas_aeruginosa', 'mycobacterium_ulcerans', 'listeria_monocytogenes', 'legionella_pneumophila']\n",
      "true pathogen(s): ['pseudomonas_aeruginosa']\n",
      "precision: 0.25\n",
      "predicting for patient 5\n",
      "{5: 15, 11: 37, 22: 1, 7: 2456, 16: 2}\n",
      "my prediction(s) for patient 5: ['corynebacterium_diphtheriae', 'listeria_monocytogenes']\n",
      "true pathogen(s): ['corynebacterium_diphtheriae']\n",
      "precision: 0.5\n",
      "predicting for patient 6\n",
      "{29: 1, 10: 1, 7: 2260, 16: 1}\n",
      "my prediction(s) for patient 6: []\n",
      "true pathogen(s): ['streptococcus_pneumoniae']\n",
      "precision: 0.0\n",
      "predicting for patient 7\n",
      "{13: 464, 5: 739, 6: 155, 12: 13, 23: 2, 7: 1254}\n",
      "my prediction(s) for patient 7: ['mycobacterium_ulcerans', 'corynebacterium_diphtheriae', 'corynebacterium_ulcerans', 'mycobacterium_tuberculosis']\n",
      "true pathogen(s): ['mycobacterium_ulcerans']\n",
      "precision: 0.25\n",
      "predicting for patient 8\n",
      "{7: 2629, 16: 2}\n",
      "my prediction(s) for patient 8: []\n",
      "true pathogen(s): ['mycobacterium_tuberculosis' 'streptococcus_pneumoniae']\n",
      "precision: 0.0\n",
      "predicting for patient 9\n",
      "{13: 4, 16: 2, 11: 119, 22: 2, 10: 9, 27: 2, 21: 2, 5: 1, 30: 1, 7: 1990}\n",
      "my prediction(s) for patient 9: ['listeria_monocytogenes', 'legionella_pneumophila']\n",
      "true pathogen(s): ['streptococcus_pneumoniae']\n",
      "precision: 0.0\n",
      "predicting for patient 10\n",
      "{1: 215, 21: 2, 11: 182, 20: 1, 5: 27, 16: 6, 13: 5, 6: 3, 10: 2, 22: 1, 0: 2, 27: 1, 7: 2126}\n",
      "my prediction(s) for patient 10: ['burkholderia_pseudomallei', 'listeria_monocytogenes', 'corynebacterium_diphtheriae', 'pseudomonas_aeruginosa']\n",
      "true pathogen(s): ['burkholderia_pseudomallei']\n",
      "precision: 0.25\n",
      "['patient 1: 0.0', 'patient 2: 0.1111111111111111', 'patient 3: 0.2857142857142857', 'patient 4: 0.25', 'patient 5: 0.5', 'patient 6: 0.0', 'patient 7: 0.25', 'patient 8: 0.0', 'patient 9: 0.0', 'patient 10: 0.25']\n",
      "avg: 0.16468253968253968\n",
      "Time taken for prediction: 4518.304705142975 seconds\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "\n",
    "def predict(prob_threshold, read_threshold, model, transformer):\n",
    "    start = time.time()\n",
    "\n",
    "    all_precision = []\n",
    "    for patient_id in range(1, 11):\n",
    "        print(f'predicting for patient {patient_id}')\n",
    "        \n",
    "        with open(f'datasets/validation/patient{patient_id}.6mer.npy', 'rb') as read_file:\n",
    "            df_test = np.load(read_file)\n",
    "        df_test = pd.DataFrame(df_test)\n",
    "        X_test = transformer.transform(df_test.loc[:, feat].values)\n",
    "\n",
    "        # we can use model.predict_proba to find a good threshold and predict only for cases where the model is confident\n",
    "        y_predprob = model.predict_proba(X_test) # probability of each class per read\n",
    "        \n",
    "        # we get only predictions larger than the threshold and if there is more than one, we take the argmax again\n",
    "        # num_reads = X_test.shape[0]\n",
    "        predictions_count = {}\n",
    "        for read in y_predprob:\n",
    "            if np.max(read) >= prob_threshold:\n",
    "                label = np.argmax(read)\n",
    "                if label in predictions_count:\n",
    "                    predictions_count[label] += 1\n",
    "                else:\n",
    "                    predictions_count[label] = 1\n",
    "        print(predictions_count)\n",
    "        final_predictions = le.inverse_transform([label for label, count in predictions_count.items() if count > read_threshold]) # only consider labels with more than 5 reads predicting that label, can change to % num_reads\n",
    "        \n",
    "        # my pathogens detected, decoy will be ignored\n",
    "        final_predictions = [item for item in final_predictions if item !='decoy']\n",
    "        \n",
    "        precision = precision_per_patient(patient_id, final_predictions)\n",
    "        print(f'precision: {precision}')\n",
    "        all_precision.append(precision)\n",
    "\n",
    "    # performance per patient and its final average\n",
    "    print([f'patient {c}: {item}' for c, item in enumerate(all_precision, start=1)])\n",
    "    print(f'avg: {np.mean(all_precision)}')\n",
    "\n",
    "    end = time.time()\n",
    "    print(f'Time taken for prediction: {end - start} seconds')\n",
    "\n",
    "predict(0.8, 5, model, svd)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
